File: apps/api/app/schemas/envelope.py
"""API response envelope schemas."""

from typing import Any

from pydantic import BaseModel, Field

from app.domain.shared.enums import (
    ApiStatus,
    ConfidenceLevel,
    PassTendencyLevel,
    RiskLevel,
    SourceAvailability,
)
from app.schemas.refs import EvidenceRefSchema, LowConfidenceFlagSchema, ResourceRef, TraceRefSchema


class ApiSuccessEnvelope(BaseModel):
    request_id: str
    trace_id: str
    status: ApiStatus
    resource_type: str
    resource_ref: ResourceRef | None = None
    data: Any | None = None
    meta: dict[str, Any] | None = None
    candidate_refs: list[ResourceRef] = Field(default_factory=list)
    suggestion_refs: list[ResourceRef] = Field(default_factory=list)
    confirmation_required: bool | None = None
    ai_task_id: str | None = None
    validation_result_ref: ResourceRef | None = None
    low_confidence_flags: list[LowConfidenceFlagSchema] = Field(default_factory=list)
    source_availability: SourceAvailability | None = None
    evidence_refs: list[EvidenceRefSchema] = Field(default_factory=list)
    trace_refs: list[TraceRefSchema] = Field(default_factory=list)
    score_version: str | None = None
    rubric_version: str | None = None
    confidence_level: ConfidenceLevel | None = None
    pass_tendency_level: PassTendencyLevel | None = None
    risk_level: RiskLevel | None = None
    next_actions: list[str] = Field(default_factory=list)


class ApiError(BaseModel):
    code: str
    message: str
    details: dict[str, Any] | None = None
    retryable: bool
    user_action: str | None = None
    audit_event_ref: ResourceRef | None = None


class ApiErrorEnvelope(BaseModel):
    request_id: str
    trace_id: str
    error: ApiError

File: apps/web/src/shared/api/envelope.ts
export interface ApiResourceRef {
  resource_type: string;
  resource_id: string;
}

export interface ApiSuccessEnvelope<TData = unknown> {
  request_id: string;
  trace_id: string;
  status: string;
  resource_type: string;
  resource_ref?: ApiResourceRef | null;
  data: TData | null;
  meta?: Record<string, unknown> | null;
  candidate_refs?: ApiResourceRef[];
  suggestion_refs?: ApiResourceRef[];
  confirmation_required?: boolean | null;
  ai_task_id?: string | null;
  validation_result_ref?: ApiResourceRef | null;
  low_confidence_flags?: unknown[];
  source_availability?: string | null;
  evidence_refs?: unknown[];
  trace_refs?: unknown[];
  score_version?: string | null;
  rubric_version?: string | null;
  confidence_level?: string | null;
  pass_tendency_level?: string | null;
  risk_level?: string | null;
  next_actions?: string[];
}

export interface ApiErrorPayload {
  code: string;
  message: string;
  details?: Record<string, unknown> | null;
  retryable: boolean;
  user_action?: string | null;
  audit_event_ref?: ApiResourceRef | null;
}

export interface ApiErrorEnvelope {
  request_id: string;
  trace_id: string;
  error: ApiErrorPayload;
}

export const DEFAULT_UNAUTHENTICATED_ERROR_CODE = "unauthenticated";
export const DEFAULT_INTERNAL_ERROR_CODE = "internal_error";
